README.md
# compass-rails

**We Take Pull Requests!**

[![Build Status](https://travis-ci.org/Compass/compass-rails.png?branch=master)](https://travis-ci.org/Compass/compass-rails)
[![Code Climate](https://codeclimate.com/github/Compass/compass-rails.png)](https://codeclimate.com/github/Compass/compass-rails)
[![Gem Version](https://badge.fury.io/rb/compass-rails.svg)](http://badge.fury.io/rb/compass-rails)
[![Coverage Status](https://coveralls.io/repos/Compass/compass-rails/badge.png)](https://coveralls.io/r/Compass/compass-rails)

Compass rails is an adapter for the [Compass Stylesheet Authoring
Framework](http://compass-style.org) for [Ruby on Rails](http://rubyonrails.org/).

Since Compass v0.12.0, this is the only way to use compass with your rails application.

Supports Rails 3.2, 4.x releases.

## Installation

Add the `compass-rails` gem line to your application's Gemfile

```ruby
gem 'sass-rails'
gem 'compass-rails'
```

If you are using any Compass extensions, add them to this group in your
Gemfile.

And then execute:

    $ bundle

## Usage

Change your `application.css` to `application.css.scss` or `application.css.sass` and then `@import compass` and your own stylesheets to your hearts content. E.g.:

```scss
@import "compass";

@import "your_project/mixins";
@import "your_project/base";
```

*or*

Use `application.css` to require files that use compass features. Ex:
```css
/*
 *= require styleguide_full_of_compass_stuff
 */
```

*Don't* use `*= require something` within your SCSS or SASS files. You're gonna have a bad time.

### Configuration

Compass-rails is configured out of the box to work with Rails. 

Advanced users can choose to add a `config/compass.rb` and take advantage of the [Compass configuration 
reference](http://compass-style.org/help/documentation/configuration-reference/)
as is.

### Installing Compass extensions

Step 1: Add it to your Gemfile and run the `bundle` command to install it.

Step 2: Install the extension's assets: `bundle exec compass install 
<extension/template>`

For example, if you want to use susy.

```ruby
# Gemfile
gem 'compass-rails'
gem 'compass-susy-plugin'
```

then run:

    $ bundle
    $ bundle exec compass install susy

if you are using the rails configuration files you should add:

```ruby
config.compass.require "susy"
```

to your application.rb configuration file.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
# Change log

## 3.0.0 - Unreleased
### Added
- Added Sprockets 3 support. Issue #232, #236, #244.

### Fixed
- Sprockets cache files are no longer saved to /tmp and now use the app-level tmp folder instead.

## 2.0.2 - 2015-01-03
### Fixed
- Fixed test suite to test against Rails 3.1, 3.2, 4.0, 4.2 on Ruby 1.9.3, 2.0.0, 2.1.0, 2.2.0 and jruby-head. Issue #206
- Support up to sass-rails 5.0.1. Issue #198
- Fix sass_importer patches having incorrect method signatures. Issue #195
- Fixed incorrect path generation for sprites in Rails 4. Issue #190

## 2.0.0 - 2014-07-10
### Added
- Support for Rails 4.2
- Allow newer sass-rails with Sass 3.3 support (see #160).

### Removed
- Rails 3.0 support.

### Fixed
- Properly bust the cache on image sprites within a sub-directory that
are imported with a wildcard (see #166).

## 1.1.7 - 2014-03-18
### Fixed
- Locked Sprockets version to 2.11.0 (see #146).

## 1.1.6 - 2014-03-11
### Fixed
- Leave bundle selection to rails environment.

## 1.1.5 - 2014-03-11
### Added
- Support for Ruby 2.1.0.

### Fixed
- Fixed `rails_loaded?` for when Rails is defined but no application is actually loaded.

## 1.1.4 - 2014-03-18
### Changed
- Simplified README.

## 1.1.3 - 2013-12-27
### Fixed
- No longer assuming asset pipeline is running when generating sprites.

## 1.1.2 - 2013-12-06
### Fixed
- Reverted fix for `generated_image_url` introduced in 1.1.0.

## 1.1.1 - 2013-12-05
### Added
- Support for Compass versions greater than 0.12.2.

## 1.1.0 - 2013-12-05
### Added
- Rails 4 support.
- Ruby 2.0 support.

### Fixed
- Allow compass-rails without asset pipeline on Rails 3.2.
- Fix `generated_image_url` when `generated_images_dir` is set.

### Removed
- Support for Ruby 1.8.7 and 1.9.2.
- Support for Rails 2.3.

## 1.0.3 - 2012-06-26
### Added
- Bumped Compass version to 0.12.2.

### Fixed
- `FixedStaticCompiler` hack so sprite source files dont show up in manifest.


lib/compass-rails.rb
require 'compass'
require "compass-rails/version"
require "compass-rails/configuration"

module CompassRails

    RAILS_32 = %r{^3.2}
    RAILS_31 = %r{^3.1}

    extend self

    def setup_fake_rails_env_paths(sprockets_env)
      return unless rails_loaded?

      if sprockets_env.respond_to?(:trail, true)
        sprockets_trail = sprockets_env.send(:trail)
      else
        sprockets_trail = sprockets_env.index
      end

      keys = ['app/assets', 'lib/assets', 'vendor/assets']
      local = keys.map {|path| ::Rails.root.join(path) }.map { |path| [File.join(path, 'images'), File.join(path, 'stylesheets')] }.flatten!
      sprockets_trail.paths.unshift(*local)
      paths = []
      ::Rails::Engine.subclasses.each do |subclass|
        paths = subclass.paths
        keys.each do |key|
          sprockets_trail.paths.unshift(*paths[key].existent_directories)
        end
      end
    end

    def sass_config
      ::Rails.application.config.sass
    end

    def sprockets
      @sprockets ||= ::Rails.application.assets
    end

    def context
      @context ||= begin
        sprockets.version = ::Rails.env + "-#{sprockets.version}"
        setup_fake_rails_env_paths(sprockets)
        context = ::Rails.application.assets.context_class
        context.extend(::Sprockets::Helpers::IsolatedHelper)
        context.extend(::Sprockets::Helpers::RailsHelper)
        context.extend(::Sass::Rails::Railtie::SassContext)
        context.sass_config = sass_config

        context
      end
    end

    def rails_loaded?
      defined?(::Rails)
    end

    def rails_version
      rails_spec = (Gem.loaded_specs["railties"] || Gem.loaded_specs["rails"])
      raise "You have to require Rails before compass" unless rails_spec
      rails_spec.version.to_s
    end

    def rails31?
      return false unless defined?(::Rails)
      version_match RAILS_31
    end

    def rails32?
      return false unless defined?(::Rails)
      version_match RAILS_32
    end

    def version_match(version)
      !(rails_version =~ version).nil?
    end

    def configuration
      config = Compass::Configuration::Data.new('rails')
      config.extend(CompassRails::Configuration)
      config
    end

    def prefix
      ::Rails.application.config.assets.prefix
    end

    def configure_rails!(app)
      return unless app.config.respond_to?(:sass)
      sass_config = app.config.sass
      compass_config = app.config.compass

      sass_config.load_paths.concat(compass_config.sass_load_paths)

      { :output_style => :style,
        :line_comments => :line_comments,
        :cache => :cache,
        :disable_warnings => :quiet,
        :preferred_syntax => :preferred_syntax
      }.each do |compass_option, sass_option|
        set_maybe sass_config, compass_config, sass_option, compass_option
      end
      if compass_config.sass_options
        compass_config.sass_options.each do |config, value|
          sass_config.send("#{config}=", value)
        end
      end
    end

  private

    # sets the sass config value only if the corresponding compass-based setting
    # has been explicitly set by the user.
    def set_maybe(sass_config, compass_config, sass_option, compass_option)
      if compass_value = compass_config.send(:"#{compass_option}_without_default")
        sass_config.send(:"#{sass_option}=", compass_value)
      end
    end

end

if defined?(::Rails)
  Compass::AppIntegration.register(:rails, "::CompassRails")
  require "compass-rails/patches"
  require "compass-rails/railties"
end

